libs/common-services/src/main/java/uk/gov/di/ipv/core/library/service/IpvSessionService.java

```diff
@@ -58,32 +66,34 @@ public Optional<IpvSessionItem> getIpvSessionByAuthorizationCode(String authoriz
     }
 
     public Optional<IpvSessionItem> getIpvSessionByAccessToken(String accessToken) {
-
         IpvSessionItem ipvSessionItem = null;
-
-        // Simple back off and retry
-        var attempts = 0;
-        while (attempts < 7) {
+        try {
             ipvSessionItem =
-                    dataStore.getItemByIndex("accessToken", DigestUtils.sha256Hex(accessToken));
-
-            if (ipvSessionItem == null) {
-                var backoff = (long) (10 * Math.pow(2, attempts++));
-                try {
-                    Thread.sleep(backoff);
-                } catch (InterruptedException e) {
-                    LOGGER.warn(
-                            LogHelper.buildLogMessage(
-                                    "getIpvSessionByAccessToken() backoff and retry sleep was interrupted"));
-                    Thread.currentThread().interrupt();
-                }
-            } else {
-                break;
-            }
-        }
-
-        if (attempts > 0) {
-            LOGGER.warn(LogHelper.buildLogMessage("getIpvSessionByAccessToken() required retries"));
+                    Retry.runTaskWithBackoff(
+                            sleeper,
+                            7,
+                            10,
+                            () -> {
+                                var item =
+                                        dataStore.getItemByIndex(
+                                                "accessToken", DigestUtils.sha256Hex(accessToken));
+                                if (item == null) {
+                                    throw new RetryableException(
+                                            new UnknownAccessTokenException(
+                                                    "Failed to find access token"));
+                                }
+                                return item;
+                            });
+        } catch (InterruptedException e) {
+            LOGGER.warn(
+                    LogHelper.buildLogMessage(
+                            "getIpvSessionByAccessToken() backoff and retry sleep was interrupted"));
+            Thread.currentThread().interrupt();
+        } catch (NonRetryableException e) {
+            LOGGER.warn(
+                    LogHelper.buildErrorMessage(
+                            "getIpvSessionByAccessToken() exception occurred retrying getItemByIndex",
+                            e));
         }
         return Optional.ofNullable(ipvSessionItem);
```

Comment:
Can we just return the retry result as a (non-optional) `IpvSession`?

And when catching the Non-Retryable exception, we need to do something similar to `EvcsClient` where we pass through a wrapped `UnknownAccessException` and wrap other errors appropriately.

Reply:
Yes, I was just trying to keep the method signature the same to minimise changes

For the exception, I think we are only logging it so is there any need to handle the different types?

Reply:
So do you mean to throw the error and let the callers handle it?

Reply:
Yeah - if you look at the calling code, it checks if it's empty and then throws its own `UnknownAccessTokenException` (which you can pull down into here)